Convert 0.0.0.0 to local ip addresses when  matching the cluster node

zevin02 opened the pull request:
fixes: #1381

- Add a function to get all local IP addresses

zevin02 committed:
feat: implement local IP address retrieval function

fixes: #1381

- Add a function to get all local IP addresses

Signed-off-by: Zewen Xu <zevin9427@gmail.com>

torwig commented on src/cluster/cluster.cc:
```diff
@@ -157,7 +159,34 @@ Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_versio
 }
+// Get all local IP addresses.
+std::vector<std::string> getLocalIPAddresses() {
+  std::vector<std::string> ipAddresses;
```
Hi @zevin02 , thank you for your contribution.
Please, rename the function's local variables like `ip_addresses`.
Also, you can use the `nullptr` keyword instead of `NULL` and simplify checks like `if (pointer != NULL)` by using `if (pointer)`

torwig commented on src/cluster/cluster.cc:
```diff
@@ -157,7 +159,34 @@ Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_versio
+      char addressBuffer[INET_ADDRSTRLEN];
+      inet_ntop(AF_INET, tmpAddrPtr, addressBuffer, INET_ADDRSTRLEN);
+      std::string ipAddress(addressBuffer);
+      ipAddresses.push_back(move(ipAddress));
```
I would add `std::` prefix to the `move` function.

git-hulk replied:
It looks unnecessary to get IP addresses, we can do it once if finds the listening address was on 0.0.0.0 when starting.

zevin02 replied:
Can I understand that if the server listens to 0.0.0.0, any cluster node with any ip on the same port needs to be matched, or what?

git-hulk replied:
No actually, we can only match the IP addresses that the current node is listening. And 0.0.0.0 is a special case since it represents listening on all IP addresses in this node, we need to convert to the real IP(like 192.168.1.4) + loopback IP(127.0.0.1) before matching.

zevin02 replied:
So is there a problem with my above operation? If it is 0.0.0.0, get all local ip and cluster node ip for matching

git-hulk replied:
The implementation of `getLocalIPAddresses` looks good. My point is that we should use all IP addresses to match only if the listening address is `0.0.0.0`.

zevin02 replied:
Got it, thanks

torwig commented:
@zevin02 It's very handy to run `./x.py format` before committing the changes - it will format the C++ source code automatically so the CI's linter will be happy.

PragmaTwice commented:
Hi @zevin02, thanks for your contribution!

We setup lots of code checks in our CI, which requires you to pass for better code quality.

For example, from a glance to your code, maybe these checks will fail: 
- local variables must be lower_case with underscore,
- NULL must be replaced with nullptr

You can check the CI log for more details and for how to modify your code to pass the CI.

It is recommended to try some commands to solve these problems, but maybe not all problems:
- `./x.py format`
- `./x.py check tidy --fix`

Another thing is that, it would be better to have some unit tests for this function before it is merged : )

enjoy-binbin commented on src/cluster/cluster.cc:
```diff
@@ -201,14 +230,20 @@ Status Cluster::SetClusterNodes(const std::string &nodes_str, int64_t version, b
     for (auto &n : nodes_) {
-      if (n.second->port == port_ && std::find(binds_.begin(), binds_.end(), n.second->host) != binds_.end()) {
-        myid_ = n.first;
+      //If the IP being listened on is 0.0.0.0, we need to compare the node's IP address with all local IP addresses.
```
```suggestion
      // If the IP being listened on is 0.0.0.0, we need to compare the node's IP address with all local IP addresses.
```

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -157,6 +159,33 @@ Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_versio
   return Status::OK();
 }
+// Get all local IP addresses.
+std::vector<std::string> GetLocalIpAddresses() {
```
I think it would be better to put this function in `common/io_util.h/cc`.

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -157,6 +159,33 @@ Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_versio
+      char address_buffer[INET_ADDRSTRLEN];
+      inet_ntop(AF_INET, tmp_addr_ptr, address_buffer, INET_ADDRSTRLEN);
+      ip_addresses.emplace_back(std::string(address_buffer));
+    }
```
It seems the IPv6 part is missing, which is supported by kvrocks in other components.

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -157,6 +159,33 @@ Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_versio
+std::vector<std::string> GetLocalIpAddresses() {
+  std::vector<std::string> ip_addresses;
+
+  struct ifaddrs *if_addr_struct = nullptr;
```
Seems the `struct` prefix is useless here.

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -157,6 +159,33 @@ Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_versio
+  std::vector<std::string> ip_addresses;
+
+  struct ifaddrs *if_addr_struct = nullptr;
+  struct ifaddrs *ifa = nullptr;
+  void *tmp_addr_ptr = nullptr;
```
Please define these variables closer to where they are used.

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -157,6 +159,33 @@ Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_versio
+      tmp_addr_ptr = &((struct sockaddr_in *)ifa->ifa_addr)->sin_addr;
+      char address_buffer[INET_ADDRSTRLEN];
+      inet_ntop(AF_INET, tmp_addr_ptr, address_buffer, INET_ADDRSTRLEN);
+      ip_addresses.emplace_back(std::string(address_buffer));
```
```suggestion
      ip_addresses.emplace_back(address_buffer);
```

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -157,6 +159,33 @@ Status Cluster::SetSlot(int slot, const std::string &node_id, int64_t new_versio
+    }
+  }
+
+  if (if_addr_struct) freeifaddrs(if_addr_struct);
```
You can try to use `std::unique_ptr` or `ScopeExit` for `if_addr_struct` to avoid manually free.

zevin02 replied:
got it

git-hulk replied:
@zevin02 can just append fix commits directly without rebasing, then reviewers can review the new changes only.

zevin02 replied:
OK, I will submit PR like this in the future

zevin02 replied:
I tried a bit using smart pointers to avoid manual freeing, it seems that there is some difficulty for me, can you give me some advice

zevin02 committed:
refactor: refactor network functions into common library

- Move the `GetLocalIpAddresses` function to `common/io_util.cc`
- Add `common/io_util.h` header to `cluster/cluster.h`
- Use `util::GetLocalIpAddresses()` instead of the previously defined function in `cluster.cc`

Signed-off-by: Zewen Xu <zevin9427@gmail.com>

zevin02 commented:
@PragmaTwice I think I solved all the problems, can you give me a review

PragmaTwice commented:
> @PragmaTwice I think I solved all the problems, can you give me a review

Sure! But as I mentioned in https://github.com/apache/incubator-kvrocks/pull/1392#issuecomment-1511603694, you need first to make the CI pass, which currently failed.

```
./x.py format
```

zevin02 committed:
Format code

Signed-off-by: Zewen Xu <zevin9427@gmail.com>

PragmaTwice committed:
Merge branch 'unstable' into fix-ipaddr

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -20,9 +20,11 @@
 
 #include "cluster.h"
 
+#include <arpa/inet.h>
 #include <config/config_util.h>
+#include <ifaddrs.h>
+#include <sys/types.h>
```
Seems these newly added headers are no longer needed in this file.

PragmaTwice commented on src/common/io_util.h:
```diff
@@ -20,7 +20,9 @@
 
+#include <ifaddrs.h>
 #include <netinet/in.h>
+#include <sys/types.h>
```
These headers can be put into `.cc`, rather than `.h`.

PragmaTwice commented on src/cluster/cluster.h:
```diff
@@ -30,6 +30,7 @@
 #include <vector>
 
 #include "commands/commander.h"
+#include "common/io_util.h"
```
ditto

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -199,16 +201,34 @@ Status Cluster::SetClusterNodes(const std::string &nodes_str, int64_t version, b
+  // Store all the local ip
+  std::unique_ptr<std::vector<std::string>> local_hosts(nullptr);
+
+  if (std::find(binds_.begin(), binds_.end(), "0.0.0.0") != binds_.end()) {
```
Seems the ipv6 address INADDR6_ANY (`::`) is missing.

PragmaTwice commented on src/cluster/cluster.cc:
```diff
@@ -199,16 +201,34 @@ Status Cluster::SetClusterNodes(const std::string &nodes_str, int64_t version, b
+  // Used to check if listening on "0.0.0.0“
+  bool is_listen_all_ip = false;
+  // Store all the local ip
+  std::unique_ptr<std::vector<std::string>> local_hosts(nullptr);
```
```suggestion
  std::unique_ptr<std::vector<std::string>> local_hosts;
```

PragmaTwice reviewed (commented):
The rest looks good to me. cc @git-hulk

git-hulk committed:
Minor refactor the match listening ip

git-hulk committed:
Fix include headers

git-hulk commented:
> The rest looks good to me. cc @git-hulk

@PragmaTwice I do a minor refactor, can help to take a look again.

torwig commented on src/common/io_util.cc:
```diff
@@ -312,6 +314,53 @@ int AeWait(int fd, int mask, int timeout) {
+  return false;
+}
+
+// Get all local IP addresses.
```
I believe the function's name is so descriptive that we can omit this comment.

torwig commented on src/common/io_util.cc:
```diff
@@ -312,6 +314,53 @@ int AeWait(int fd, int mask, int timeout) {
+std::vector<std::string> GetLocalIPAddresses() {
+  std::vector<std::string> ip_addresses;
+  ifaddrs *if_addr_struct = nullptr;
+  // Use unique_ptr for if_addr_struct to avoid manually free
```
Also here we can omit the comment in my opinion.

torwig commented on src/common/io_util.cc:
```diff
@@ -312,6 +314,53 @@ int AeWait(int fd, int mask, int timeout) {
+    }
+    void *tmp_addr_ptr = nullptr;
+    if (ifa->ifa_addr->sa_family == AF_INET) {
+      // check it is IP4 and not a loopback address
```
The comment states that there is a check that it's not a loopback address, but I don't see any checks. Is everything OK here?
The same for the `else` branch.

zevin02 replied:
I rechecked this code, I didn't check the local loopback address, I think it's ok, so I should delete the extra comment after that, right? or should I check if it's the local loopback address?

git-hulk replied:
Yes, the comment should be removed if we didn't check that. It looks no necessary to check if it's a loopback address.

PragmaTwice replied:
The comment should be what the code does, otherwise it must be `TODO` or `FIXME`.

torwig reviewed (commented):
Generally, LGTM, but I have a few comments.

zevin02 committed:
refactor: refactor network address functions

- Check for IPv6 addresses in `MatchListeningIP`
- Remove unnecessary comment about loopback addresses in `GetLocalIPAddresses`

Signed-off-by: Zewen Xu <zevin9427@gmail.com>

torwig reviewed (approved):
LGTM

PragmaTwice commented:
Thanks for your contribution! Merging...
